DOC: Add whatsnew illustrating upcoming changes (#56545)
pandas-dev/pandas: doc/source/whatsnew/v2.2.0.rst

{{ header }}

.. ---------------------------------------------------------------------------

> .. _whatsnew_220.upccoming_changes:

Reviewer: ```suggestion
.. _whatsnew_220.upcoming_changes:
```

Upcoming changes in pandas 3.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

pandas 3.0 will bring two bigger changes to the default behavior of pandas.

Copy-on-Write
^^^^^^^^^^^^^

The currently optional mode Copy-on-Write will be enabled by default in pandas 3.0. There
won't be an option to keep the current behavior enabled. The new behavioral semantics are
explained in the :ref:`user guide about Copy-on-Write <copy_on_write>`.

The new behavior can be enabled since pandas 2.0 with the following option:

.. code-block:: ipython

   pd.options.mode.copy_on_write = True

This change brings different changes in behavior in how pandas operates with respect to
copies and views. Some of these changes allow a clear deprecation, like the changes in
chained assignment. Other changes are more subtle and thus, the warnings are hidden behind
an option that can be enabled in pandas 2.2.

.. code-block:: ipython

   pd.options.mode.copy_on_write = "warn"

This mode will warn in many different scenarios that aren't actually relevant to
most queries. We recommend exploring this mode, but it is not necessary to get rid
of all of these warnings.

> 

Reviewer: Can you end here with pointing to the "migration guide"?

Author: added

> Arrow-backed strings by default

Reviewer: ```suggestion
Dedicated string data type (backed by Arrow) by default
```

What do you think of such a change? From a user point of view, I think one of the important changes (apart from performance) is also that it now just an actual _string_ data type, instead of a catch-all object dtype

Author: Yeah makes sense

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Historically, pandas represented string columns with NumPy object data type. This
representation has numerous problems, including slow performance and a large memory
footprint. This will change in pandas 3.0. pandas will start inferring string columns

> as Arrow backed strings, which represents strings contiguous in memory. This brings

Reviewer: ```suggestion
as a new ``string`` data type, backed by Arrow, which represents strings contiguous in memory. This brings
```

a huge performance and memory improvement.

Old behavior:

.. code-block:: ipython

    In [1]: ser = pd.Series(["a", "b"])
    Out[1]:
    0    a
    1    b
    dtype: object

New behavior:

>     In [1]: ser = pd.Series(["a", "b"])

Reviewer: ```suggestion

.. code-block:: ipython

    In [1]: ser = pd.Series(["a", "b"])
```

    Out[1]:
    0    a
    1    b

>     dtype: string

Reviewer 2: ```suggestion
    dtype: string[pyarrow]
```
?

Author: The repr is only string, and that’s also what happens in 3.0, that’s why I kept kt

Reviewer 2: Ah ok

The string data type that is used in these scenarios will mostly behave as NumPy

> object would, this includes missing value semantics and general operations on these

Reviewer: ```suggestion
object would, including missing value semantics and general operations on these
```

columns.

> This change include a few additional changes across the API:

Reviewer: ```suggestion
This change includes a few additional changes across the API:
```

> - Specifying ``dtype="string"`` creates a dtype that is backed by Python strings

Reviewer: ```suggestion
- Currently, specifying ``dtype="string"`` creates a dtype that is backed by Python strings
```
